File: src/main/java/com/bombproj/controller/ProjectDatabaseController.java
package com.bombproj.controller;

import com.bombproj.dto.ProjectDataTableDto;
import com.bombproj.dto.ProjectDatabaseDto;
import com.bombproj.framework.JsonResult;
import com.bombproj.framework.SessionConfig;
import com.bombproj.framework.interceptor.AuthPassport;
import com.bombproj.service.ProjectDatabaseService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/project")
public class ProjectDatabaseController {

    @Autowired
    private ProjectDatabaseService projectDatabaseService;

    @AuthPassport
    @GetMapping("/field-types")
    public JsonResult getFieldTypesByDbType(@RequestParam("dbType") String dbType) {
        return JsonResult.success(this.projectDatabaseService.getFieldTypes(dbType));
    }

    @AuthPassport
    @GetMapping("/field-indexes")
    public JsonResult getIndexes(@RequestParam("dbType") String dbType) {
        return JsonResult.success(this.projectDatabaseService.getFieldIndexes(dbType));
    }

    /**
     * 查询指定项目中的数据库列表
     *
     * @param projectId    项目 ID
     * @param databaseName 查询条件：数据库名
     */
    @AuthPassport
    @GetMapping("/{projectId}/db")
    public JsonResult pageListDatabase(@PathVariable("projectId") String projectId,
                                       @RequestParam(value = "databaseName", required = false) String databaseName) {
        return JsonResult.success(this.projectDatabaseService.listDatabases(projectId, databaseName));
    }

    /**
     * 新增数据库
     *
     * @param projectId 项目 ID
     * @param dto       请求参数，包含了数据库信息
     */
    @AuthPassport
    @PostMapping("/{projectId}/db")
    public JsonResult addDatabase(@PathVariable("projectId") String projectId,
                                  @ModelAttribute ProjectDatabaseDto dto) {
        SessionConfig sessionConfig = SessionConfig.current();
        dto.setProjectId(projectId);
        dto.setUserId(sessionConfig.getUserId());
        this.projectDatabaseService.addDatabase(dto);
        return JsonResult.success();
    }

    /**
     * 更新数据库信息
     *
     * @param projectId 项目 ID
     * @param dbId 数据库 ID
     * @param dto 要更新的数据库信息
     */
    @AuthPassport
    @PutMapping("/{projectId}/db/{dbId}")
    public JsonResult updateDatabase(@PathVariable("projectId") String projectId,
                                     @PathVariable("dbId") String dbId,
                                     @ModelAttribute ProjectDatabaseDto dto) {
        SessionConfig sessionConfig = SessionConfig.current();
        dto.setUserId(sessionConfig.getUserId());
        dto.setProjectId(projectId);
        dto.setId(dbId);
        this.projectDatabaseService.updateDatabase(dto);
        return JsonResult.success();
    }

    @AuthPassport
    @DeleteMapping("/{projectId}/db/{dbId}")
    public JsonResult deleteDatabase(@PathVariable("projectId") String projectId,
                                     @PathVariable("dbId") String dbId) {
        SessionConfig sessionConfig = SessionConfig.current();
        ProjectDatabaseDto dto = new ProjectDatabaseDto();
        dto.setProjectId(projectId);
        dto.setId(dbId);
        dto.setUserId(sessionConfig.getUserId());
        this.projectDatabaseService.deleteDatabase(dto);
        return JsonResult.success();
    }

    @AuthPassport
    @GetMapping("/{projectId}/db/{dbId}/table")
    public JsonResult pageListTables(@PathVariable("projectId") String projectId,
                                     @PathVariable("dbId") String dbId,
                                     @RequestParam(value = "tableName", required = false) String tableName,
                                     @RequestParam(value = "page", defaultValue = "1") Integer page) {
        return JsonResult.success(this.projectDatabaseService.pageListTables(tableName, dbId, projectId, page));
    }

    @AuthPassport
    @PostMapping("/{projectId}/db/{dbId}/table")
    public JsonResult addTable(@PathVariable("projectId") String projectId,
                               @PathVariable("dbId") Long dbId,
                               @ModelAttribute ProjectDataTableDto dto) {
        SessionConfig sessionConfig = SessionConfig.current();
        dto.setProjectId(projectId);
        dto.setUserId(sessionConfig.getUserId());
        dto.setDatabaseId(dbId);
        this.projectDatabaseService.addTable(dto);
        return JsonResult.success();
    }

    @AuthPassport
    @PutMapping("/{projectId}/db/{dbId}/table/{tableId}")
    public JsonResult updateTable(@PathVariable("projectId") String projectId,
                                  @PathVariable("dbId") Long dbId,
                                  @PathVariable("tableId") String tableId,
                                  @ModelAttribute ProjectDataTableDto dto) {
        SessionConfig sessionConfig = SessionConfig.current();
        dto.setUserId(sessionConfig.getUserId());
        dto.setProjectId(projectId);
        dto.setTableId(tableId);
        dto.setDatabaseId(dbId);
        this.projectDatabaseService.updateTable(dto);
        return JsonResult.success();
    }

    @AuthPassport
    @DeleteMapping("/{projectId}/db/{dbId}/table/{tableId}")
    public JsonResult deleteTable(@PathVariable("projectId") String projectId,
                                  @PathVariable("dbId") Long dbId,
                                  @PathVariable("tableId") String tableId) {
        SessionConfig sessionConfig = SessionConfig.current();
        ProjectDataTableDto dto = new ProjectDataTableDto();
        dto.setProjectId(projectId);
        dto.setTableId(tableId);
        dto.setUserId(sessionConfig.getUserId());
        dto.setDatabaseId(dbId);
        this.projectDatabaseService.deleteTable(dto);
        return JsonResult.success();
    }

    @AuthPassport
    @GetMapping("/{projectId}/db/{dbId}/table/{tableId}/indexes")
    public JsonResult getTableIndexes(@PathVariable("projectId") String projectId,
                                      @PathVariable("dbId") Long dbId,
                                      @PathVariable("tableId") String tableId,
                                      @RequestParam(value = "version", required = false) String version) {
        SessionConfig sessionConfig = SessionConfig.current();
        // TODO check permission
        return JsonResult.success(this.projectDatabaseService.getTableIndexes(tableId, projectId, version));
    }

    @AuthPassport
    @GetMapping("/{projectId}/db/{dbId}/table/{tableId}/fields")
    public JsonResult getTableFields(@PathVariable("projectId") String projectId,
                                     @PathVariable("dbId") Long dbId,
                                     @PathVariable("tableId") String tableId,
                                     @RequestParam(value = "version", required = false) String version) {
        SessionConfig sessionConfig = SessionConfig.current();
        // TODO check permission
        return JsonResult.success(this.projectDatabaseService.getTableFields(tableId, projectId, version));
    }

    @AuthPassport
    @PostMapping("/{projectId}/db/{dbId}/datatable/{tableId}/fields")
    public JsonResult addTableFields(@PathVariable("projectId") String projectId,
                                     @PathVariable("dbId") Long dbId,
                                     @PathVariable("tableId") String tableId,
                                     @RequestParam("fields") String fields) {
        SessionConfig sessionConfig = SessionConfig.current();
        this.projectDatabaseService.addOrUpdateTableFields(tableId, projectId, sessionConfig.getUserId(), dbId, fields);
        return JsonResult.success();
    }

}

File: src/main/scala/com/bombproj/controller/ProjectDatabaseController.scala
package com.bombproj.controller

import com.bombproj.dto.{ProjectDataTableDto, ProjectDatabaseDto}
import com.bombproj.framework.{JsonResult, SessionConfig}
import com.bombproj.framework.interceptor.AuthPassport
import com.bombproj.service.ProjectDatabaseService
import javax.annotation.Resource
import org.springframework.web.bind.annotation.{DeleteMapping, GetMapping, ModelAttribute, PathVariable, PostMapping, PutMapping, RequestMapping, RequestParam, RestController}

@RestController
@RequestMapping(Array("/api/project"))
class ProjectDatabaseController @Resource()(projectDatabaseService: ProjectDatabaseService) {

    @AuthPassport
    @GetMapping(Array("/field-types"))
    def getFieldTypesByDbType(@RequestParam("dbType") dbType: String): JsonResult =
        JsonResult.success(data = this.projectDatabaseService.getFieldTypes(dbType))

    @AuthPassport
    @GetMapping(Array("/field-indexes"))
    def getIndexes(@RequestParam("dbType") dbType: String): JsonResult =
        JsonResult.success(data = this.projectDatabaseService.getFieldIndexes(dbType))

    /**
     * 查询指定项目中的数据库列表
     *
     * @param projectId    项目 ID
     * @param databaseName 查询条件：数据库名
     */
    @AuthPassport
    @GetMapping(Array("/{projectId}/db"))
    def pageListDatabase(@PathVariable("projectId") projectId: String,
                         @RequestParam(value = "databaseName", required = false) databaseName: String): JsonResult =
        JsonResult.success(data = this.projectDatabaseService.listDatabases(projectId, databaseName))

    /**
     * 新增数据库
     *
     * @param projectId 项目 ID
     * @param dto       请求参数，包含了数据库信息
     */
    @AuthPassport
    @PostMapping(Array("/{projectId}/db"))
    def addDatabase(@PathVariable("projectId") projectId: String, @ModelAttribute dto: ProjectDatabaseDto): JsonResult = {
        val sessionConfig = SessionConfig.current()
        dto.setProjectId(projectId)
        dto.setUserId(sessionConfig.userId)
        this.projectDatabaseService.addDatabase(dto)
        JsonResult.success()
    }

    /**
     * 更新数据库信息
     *
     * @param projectId 项目 ID
     * @param dbId      数据库 ID
     * @param dto       要更新的数据库信息
     */
    @AuthPassport
    @PutMapping(Array("/{projectId}/db/{dbId}"))
    def updateDatabase(@PathVariable("projectId") projectId: String,
                       @PathVariable("dbId") dbId: String,
                       @ModelAttribute dto: ProjectDatabaseDto): JsonResult = {
        val sessionConfig = SessionConfig.current()
        dto.setUserId(sessionConfig.userId)
        dto.setProjectId(projectId)
        dto.setId(dbId)
        this.projectDatabaseService.updateDatabase(dto)
        JsonResult.success()
    }

    /**
     * 删除指定的数据库，同时也会删除数据库中的表信息
     *
     * @param projectId 项目 ID
     * @param dbId      数据库 ID
     */
    @AuthPassport
    @DeleteMapping(Array("/{projectId}/db/{dbId}"))
    def deleteDatabase(@PathVariable("projectId") projectId: String,
                       @PathVariable("dbId") dbId: String): JsonResult = {
        val sessionConfig = SessionConfig.current()
        val dto = new ProjectDatabaseDto
        dto.setProjectId(projectId)
        dto.setId(dbId)
        dto.setUserId(sessionConfig.userId)
        this.projectDatabaseService.deleteDatabase(dto)
        JsonResult.success()
    }

    /**
     * 获取指定数据库的所有表信息
     *
     * @param projectId 项目 ID
     * @param dbId      数据库 ID
     * @param tableName 表名
     * @param page      页码
     */
    @AuthPassport
    @GetMapping(Array("/{projectId}/db/{dbId}/table"))
    def pageListTables(@PathVariable("projectId") projectId: String,
                       @PathVariable("dbId") dbId: String,
                       @RequestParam(value = "tableName", required = false) tableName: String,
                       @RequestParam(value = "page", defaultValue = "1") page: Integer): JsonResult =
        JsonResult.success(data = this.projectDatabaseService.pageListTables(tableName, dbId, projectId, page))

    /**
     * 在指定的数据库中新增表
     *
     * @param projectId 项目 ID
     * @param dbId      数据库 ID
     * @param dto       新增的表的信息
     */
    @AuthPassport
    @PostMapping(Array("/{projectId}/db/{dbId}/table"))
    def addTable(@PathVariable("projectId") projectId: String,
                 @PathVariable("dbId") dbId: Long,
                 @ModelAttribute dto: ProjectDataTableDto): JsonResult = {
        val sessionConfig = SessionConfig.current()
        dto.setProjectId(projectId)
        dto.setUserId(sessionConfig.userId)
        dto.setDatabaseId(dbId)
        JsonResult.success(data = this.projectDatabaseService.addTable(dto))
    }

    /**
     * 更新表信息，包括表名、列信息、索引信息
     *
     * @param projectId 项目 ID
     * @param dbId      数据库 ID
     * @param tableId   表 ID
     * @param dto
     */
    @AuthPassport
    @PutMapping(Array("/{projectId}/db/{dbId}/table/{tableId}"))
    def updateTable(@PathVariable("projectId") projectId: String,
                    @PathVariable("dbId") dbId: Long,
                    @PathVariable("tableId") tableId: String,
                    @ModelAttribute dto: ProjectDataTableDto): JsonResult = {
        val sessionConfig = SessionConfig.current()
        dto.setUserId(sessionConfig.userId)
        dto.setProjectId(projectId)
        dto.setTableId(tableId)
        dto.setDatabaseId(dbId)
        this.projectDatabaseService.updateTable(dto)
        JsonResult.success()
    }

    /**
     * 删除指定的表
     *
     * @param projectId 项目 ID
     * @param dbId      数据库 ID
     * @param tableId   要删除的表 ID
     */
    @AuthPassport
    @DeleteMapping(Array("/{projectId}/db/{dbId}/table/{tableId}"))
    def deleteTable(@PathVariable("projectId") projectId: String,
                    @PathVariable("dbId") dbId: Long,
                    @PathVariable("tableId") tableId: String): JsonResult = {
        val sessionConfig = SessionConfig.current()
        val dto = new ProjectDataTableDto
        dto.setProjectId(projectId)
        dto.setTableId(tableId)
        dto.setUserId(sessionConfig.userId)
        dto.setDatabaseId(dbId)
        this.projectDatabaseService.deleteTable(dto)
        JsonResult.success()
    }

    /**
     * 获取指定表的索引信息
     *
     * @param projectId 项目 ID
     * @param dbId      数据库 ID
     * @param tableId   表 ID
     * @param version   版本号
     */
    @AuthPassport
    @GetMapping(Array("/{projectId}/db/{dbId}/table/{tableId}/indexes"))
    def getTableIndexes(@PathVariable("projectId") projectId: String,
                        @PathVariable("dbId") dbId: Long,
                        @PathVariable("tableId") tableId: String,
                        @RequestParam(value = "version", required = false) version: String): JsonResult = {
        val sessionConfig = SessionConfig.current()
        // TODO check permission
        JsonResult.success(data = this.projectDatabaseService.getTableIndexes(tableId, projectId, version))
    }

    /**
     * 获取指定表的字段信息
     *
     * @param projectId 项目 ID
     * @param dbId      数据库 ID
     * @param tableId   表 ID
     * @param version   版本号
     */
    @AuthPassport
    @GetMapping(Array("/{projectId}/db/{dbId}/table/{tableId}/fields"))
    def getTableFields(@PathVariable("projectId") projectId: String,
                       @PathVariable("dbId") dbId: Long,
                       @PathVariable("tableId") tableId: String,
                       @RequestParam(value = "version", required = false) version: String): JsonResult =
        JsonResult.success(data = this.projectDatabaseService.getTableFields(tableId, projectId, version))

    @AuthPassport
    @PostMapping(Array("/{projectId}/db/{dbId}/datatable/{tableId}/fields"))
    def addTableFields(@PathVariable("projectId") projectId: String,
                       @PathVariable("dbId") dbId: Long,
                       @PathVariable("tableId") tableId: String,
                       @RequestParam("fields") fields: String): JsonResult = {
        val sessionConfig = SessionConfig.current()
        this.projectDatabaseService.addOrUpdateTableFields(tableId, projectId, sessionConfig.userId, dbId, fields)
        JsonResult.success()
    }

}
